lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,402 @@
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing lapse rate calculation plugins."""
+
+import iris
+import numpy as np
+from numpy.linalg import lstsq
+from improver.utilities.cube_manipulation import enforce_float32_precision
+from improver.constants import DALR
+
+from scipy.ndimage import generic_filter
+
+
+class GenericFunc(object):
+    """
+    The "generic_filter" function extracts the neighbourhood around
+    each point as "buffer". This saves the buffer into "allbuffers"
+    array for the next processing step.
+    """
+    def __init__(self, allbuffers):
+        """Initialise the class.
+
+           Create the global variables that allows the "filter" function
+           to save each extracted buffer into "allbuffers".
+        """
+        # Initialises the iterator:
+        self.i = 0
+        # Saves all the buffers into this array.
+        self.allbuffers = allbuffers
+
+    def filter(self, buffer):
+        """
+        Saves the contents of the buffer into "allbuffers" array so
+        a return value isn't required. However "generic_filter"
+        requires a return value - so use zero.
+
+        Args:
+            buffer (array):
+                Array containing neighourbood points.
+
+        Returns:
+            zero (float)
+                Blank return value required by "generic_filter".
+        """
+        self.allbuffers[self.i, :] = buffer
+        self.i += 1
+        return 0.0
+
+
+class LapseRate(object):
+    """Plugin to calculate the lapse rate from orography and temperature
+       cubes.
+
+    Science background:
+
+    References:
+        The method applied here is based on the method used in the 2010 paper:
+        https://rmets.onlinelibrary.wiley.com/doi/abs/10.1002/met.177
+
+    The current lapse rate (LR) FORTRAN code is kept here:
+    http://fcm9/projects/PostProc/browser/PostProc/trunk/
+                                         model_processing/Fss_CalcTLev1Grad.f90
+
+    However the paper doesn't give critical information on certain aspects and
+    they will need more investigation - but for now its important to just have
+    a functional script.
+
+    1) The paper imposes an upper limit to the LR of 0.0098K/m-1
+       (the Dry Adibatic Lapse Rate) but not a lower limit.
+
+       Wheras the current FORTRAN code places a lower limit on the LR of
+       -3.0*DALR in addition to the upper limit.
```

Comment:
This isn't really the place to comment on "further work".  Please could you move these points (that need further investigation) to the JIRA ticket, where we can discuss and open follow on work as appropriate.

Reply:
Good idea. Done.